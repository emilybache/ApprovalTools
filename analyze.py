import argparse
import os
import re
import io
import difflib
from pprint import pprint


class DiffGroup:
    def __init__(self, diff, test_names=None):
        self.diff = diff
        if test_names:
            self.test_names = set(test_names)
        else:
            self.test_names = set()

    def add_test(self, test_name):
        self.test_names.add(test_name)

    def __repr__(self):
        return repr(sorted(self.test_names))


def analyze_diffs(failures):
    diffs = {}
    for test_name, diff in failures.items():
        diffs[test_name] = diff

    groups = analyze_groups(diffs)
    return groups


def analyze_groups(diffs):
    groups = {}
    for test_name1, diff1 in diffs.items():
        for test_name2, diff2 in diffs.items():
            if test_name1 == test_name2:
                continue
            # the diffs must be identical, not merely overlapping. This gives the test operator the best information
            if diff1 == diff2:
                if diff1 not in groups.keys():
                    groups[diff1] = DiffGroup(diff1)
                groups[diff1].add_test(test_name1)
                groups[diff1].add_test(test_name2)

    return groups


def report_diffs(diff_groups):
    result = ""
    result += "Failures can be grouped.\n"
    group_count = 0
    for group_name, group in diff_groups.items():
        result += f"Group #{group_count+1}:\n"
        result += "\n".join(sorted(group.test_names))
        result += f"\nAll share this diff:\n"
        result += group.diff
        result += "\n"
        group_count += 1

    return result


def report_failures(failures):
    result = "Failed tests:\n"
    for test_name, diff in failures.items():
        result += f"{test_name}\n"

    groups = analyze_diffs(failures)
    if groups:
        result += report_diffs(groups)

    result += "\n-----------------\n"
    result += f"Total test failure count: {len(failures)}"
    return result


def analyze(directory):
    regex = re.compile(r"(.*)\.received(\..*)")
    failures = {}
    for root, dirs, files in os.walk(directory):
        for received_filename in files:
            matches = re.findall(regex, received_filename)
            if matches:
                test_name = matches[0][0]

                received_file = str(os.path.join(root, received_filename))
                with open(received_file, encoding="utf-8") as f:
                    received_text = f.read()

                approved_filename = test_name + ".approved" + matches[0][1]
                approved_file = str(os.path.join(root, approved_filename))
                if os.path.exists(approved_file):
                    with open(approved_file, encoding="utf-8") as f:
                        approved_text = f.read()
                else:
                    approved_text = ""
                failures[test_name] = create_diff(received_text, approved_text)

    if not failures:
        return f"No failing tests found."
    else:
        return report_failures(failures)


def create_diff(received_text, approved_text):
    differ = difflib.Differ()
    diff = "\n".join(differ.compare(approved_text.splitlines(),
                                    received_text.splitlines()))
    return diff


if __name__ == "__main__":
    import sys
    parser = argparse.ArgumentParser()
    parser.add_argument("directory", help="the directory where the test results are (it will also search subdirectories)")

    args = parser.parse_args()
    directory = args.directory or os.getcwd()

    if not os.path.exists(directory):
        print(f"directory not found {directory}")
        sys.exit(-1)

    print(f"analyzing test results found in folder {directory}")
    analyze(directory)
