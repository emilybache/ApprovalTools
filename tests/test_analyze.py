import os

from approvaltests import verify

from analyze import analyze, analyze_groups, DiffGroup, report_diffs, create_diff, identical


def test_empty_dir(tmpdir):
    analysis = analyze(tmpdir)
    verify(analysis)


def test_only_approved_files(tmpdir):
    approved_file = os.path.join(tmpdir, "a.approved.txt")
    with open(approved_file, "w") as f:
        f.write("foo")
    analysis = analyze(tmpdir)
    verify(analysis)


def test_one_failure(tmpdir):
    test_name = "a"
    write_received_file(tmpdir, test_name, "foo")
    write_approved_file(tmpdir, test_name, "bar")
    analysis = analyze(tmpdir)
    verify(analysis)


def test_two_identical_failures(tmpdir):
    test_name = "a"
    write_received_file(tmpdir, test_name, "foo\n")
    write_approved_file(tmpdir, test_name, "bar\n")
    test_name = "b"
    write_received_file(tmpdir, test_name, "foo\n")
    write_approved_file(tmpdir, test_name, "bar\n")
    analysis = analyze(tmpdir)
    verify(analysis)


def test_missing_approved_file(tmpdir):
    test_name = "a"
    write_received_file(tmpdir, test_name, "foo")
    analysis = analyze(tmpdir)
    verify(analysis)


def write_received_file(tmpdir, test_name, file_contents):
    approval_file_type = "received"
    return write_approval_file(tmpdir, test_name, approval_file_type, file_contents)


def write_approved_file(tmpdir, test_name, file_contents):
    approval_file_type = "approved"
    return write_approval_file(tmpdir, test_name, approval_file_type, file_contents)


def write_approval_file(tmpdir, test_name, approval_file_type, file_contents):
    received_file = os.path.join(tmpdir, test_name + "." + approval_file_type + ".txt")
    with open(received_file, "w") as f:
        f.write(file_contents)
    return received_file


def test_find_diff_single_lines():
    failure = create_diff("foo\n".splitlines(True), "bar\n".splitlines(True))
    verify(failure)


def test_find_diff_multi_lines():
    failure = create_diff("foo\nbar\nbaz\n".splitlines(True), "bar\n".splitlines(True))
    verify(failure)


def test_groups_two_tests_same_diff():
    diff1 = """\
- bar
+ foo
"""
    diffs = {"a": diff1, "b": diff1}
    verify(str(analyze_groups(diffs, identical)))

def test_groups_two_tests_different_diff():
    diff1 = """\
- bar
+ foo
"""
    diff2 = """\
- baz
"""

    diffs = {"a": diff1, "b": diff2}
    verify(str(analyze_groups(diffs, identical)))


def test_groups_larger_number_of_tests():
    diff1 = """\
- bar
+ foo
"""
    diff2 = """\
- baz
"""

    diffs = {"a": diff1, "b": diff2, "c": diff2, "d": diff2, "e": diff1}
    verify(str(analyze_groups(diffs, identical)))


def test_overlapping_diffs():
    diff1 = """\
- bar
+ foo
"""
    diff2 = """\
+ foo
"""

    diffs = {"a": diff1, "b": diff2}
    verify(str(analyze_groups(diffs, identical)))


def test_multiple_overlapping_diffs():
    diff1 = """\
- bar
+ foo
"""
    diff2 = """\
+ foo
"""
    diff3 = """\
- bar
"""
    diffs = {"a": diff1, "b": diff2, "c": diff3}
    verify(str(analyze_groups(diffs, identical)))


def test_report_diffs():
    diff1 = '+ foo\n'
    diff2 = '- bar\n'
    diffs = {diff1: DiffGroup(diff1, ['a', 'b']), diff2: DiffGroup(diff2, ['a', 'c'])}
    verify(report_diffs(diffs))
