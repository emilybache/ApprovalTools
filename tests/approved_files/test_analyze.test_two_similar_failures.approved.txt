Failed tests:
a
b

-----------------
Additionally, tests in this list have similarities in their diffs
Group #1 (2 tests):
a
b
They have added lines:
['+ foo\n']

-----------------
Total test failure count: 2